Tests/FluentAssertions.Specs/Specialized/TaskAssertionSpecs.cs

```diff
@@ -185,6 +187,272 @@ public async Task When_task_completes_late_it_should_succeed()
+                "Expected a <System.InvalidOperationException> to be thrown,"
+                + " but found <System.NotSupportedException>:*with message \"foo\"*");
+        }
+
+        [Fact]
+        public async Task When_task_completes_without_exception_it_should_fail()
+        {
+            // Arrange
+            var timer = new FakeClock();
+            var taskFactory = new TaskCompletionSource<bool>();
+
+            // Act
+            Func<Task> action = () =>
+            {
+                return taskFactory
+                    .Awaiting(t => (Task)t.Task)
+                    .Should(timer).ThrowAsync<InvalidOperationException>();
+            };
+            taskFactory.SetResult(true);
+            timer.Complete();
+
+            // Assert
+            await action.Should().ThrowAsync<XunitException>().WithMessage(
+                "Expected a <System.InvalidOperationException> to be thrown, but no exception was thrown.");
+        }
+    }
+
+    public class ThrowWithinAsync
+    {
+        [Fact]
+        public async Task When_subject_is_null_it_should_throw()
+        {
+            // Arrange
+            Func<Task> action = null;
+
+            // Act
+            Func<Task> testAction = () => action.Should().ThrowWithinAsync<InvalidOperationException>(
+                100.Milliseconds(), "because we want to test the failure {0}", "message");
+
+            // Assert
+            await testAction.Should().ThrowAsync<XunitException>()
+                .WithMessage("*because we want to test the failure message*found <null>*");
+        }
+
+        [Fact]
+        public async Task When_subject_is_null_in_assertion_scope_it_should_throw()
+        {
+            // Arrange
+            Func<Task> action = null;
+
+            // Act
+            Func<Task> testAction = async () =>
+            {
+                using var _ = new AssertionScope();
+                await action.Should().ThrowWithinAsync<InvalidOperationException>(
+                    100.Milliseconds(), "because we want to test the failure {0}", "message");
+            };
+
+            // Assert
+            await testAction.Should().ThrowAsync<XunitException>()
+                .WithMessage("*because we want to test the failure message*found <null>*");
+        }
+
+        [Fact]
+        public async Task When_task_throws_fast_it_should_succeed()
+        {
+            // Arrange
+            var timer = new FakeClock();
+            var taskFactory = new TaskCompletionSource<bool>();
+
+            // Act
+            Func<Task> action = () =>
+            {
+                return taskFactory
+                    .Awaiting(t => (Task)t.Task)
+                    .Should(timer).ThrowWithinAsync<InvalidOperationException>(100.Milliseconds());
+            };
+            taskFactory.SetException(new InvalidOperationException("foo"));
+
+            // Assert
+            await action.Should().CompleteWithinAsync(1.Seconds());
+        }
```

Comment:
To exercise the edge cases around the `timeSpan` limit.

```cs
[Theory]
[InlineData(99)]
[InlineData(100)]
public async Task When_task_throws_fast_it_should_succeed(int delay)
{
    // Arrange
    var timer = new FakeClock();
    var taskFactory = new TaskCompletionSource<bool>();

    // Act
    Func<Task> action = () =>
    {
        return taskFactory
            .Awaiting(t => (Task)t.Task)
            .Should(timer).ThrowWithinAsync<InvalidOperationException>(100.Ticks());
    };

    timer.Delay(delay.Ticks());
    taskFactory.SetException(new InvalidOperationException("foo"));

    // Assert
    await action.Should().NotThrowAsync();
}

[Fact]
public async Task When_task_throws_slow_it_should_fail()
{
    // Arrange
    var timer = new FakeClock();
    var taskFactory = new TaskCompletionSource<bool>();

    // Act
    Func<Task> action = () =>
    {
        return taskFactory
            .Awaiting(t => (Task)t.Task)
            .Should(timer).ThrowWithinAsync<InvalidOperationException>(100.Ticks());
    };

    timer.Delay(101.Ticks());
    taskFactory.SetException(new InvalidOperationException("foo"));
    timer.Complete();

    // Assert
    await action.Should().ThrowAsync<XunitException>();
}
```

Reply:
Great work!
The only thing I've changed is using the assertions with timeout.
This is reason why I've introduced them: The tested code may cause deadlocks. In this case the test may never complete.
Asynchronous methods may always be tested with timeout.